internal/pkg/levee/levee.go

```diff
@@ -79,16 +102,114 @@ func run(pass *analysis.Pass) (interface{}, error) {
+	// rather we are interested in the points-to sets of the values in the slice.
+	vs := callValues(c)
+	if len(vs) == 0 {
+		return false
+	}
+
+	for _, v := range vs {
+		// If we ask the pointer package to analyze a value it can't analyze (i.e., not a pointer or
+		// pointer-like type), the analysis will fail and produce error output in stderr.
+		if !pointer.CanPoint(v.Type()) {
+			continue
+		}
+		// Let the configuration know that we are interested in computing this value's points-to set.
+		// The points-to set is the set of values that this pointer value might point to.
+		pointerConf.AddQuery(v)
+	}
+
+	// Perform the actual analysis, in accordance with our configuration.
+	// The pointer package's documentation says that "an error can occur only due to an internal bug".
+	// However, failing to provide a main package, or querying for a non-queryable value, will both lead
+	// to errors. We prevent these cases with explicit checks above.
+	result, err := pointer.Analyze(pointerConf)
+	if err != nil {
+		return false
+	}
+
+	// A call was analyzed successfully if performing pointer analysis on it led to a report.
+	analyzedSuccesfully = false
+	for _, ptr := range result.Queries {
+		// Obtain the points-to set for this pointer. Labels are used to get at the actual ssa.Values
+		// that are pointed to.
+		labels := ptr.PointsTo().Labels()
+		if len(labels) == 0 {
+			continue
+		}
+		for _, lab := range labels {
+			v := lab.Value()
+			switch t := v.(type) {
+			case *ssa.MakeInterface:
+				v = t.X
+			}
+			if !analysisConf.IsSource(utils.Dereference(v.Type())) {
+				continue
+			}
+			analyzedSuccesfully = true
+			reportAtSource(pass, v.(ssa.Node), (ssa.Node)(c))
+		}
+	}
+	return analyzedSuccesfully
+}
+
 func makeCall(c *ssa.Call) call.Call {
 	if sinkVarargs := varargs.New(c); sinkVarargs != nil {
 		return sinkVarargs
 	}
 	return call.Regular(c)
 }
 
-func report(pass *analysis.Pass, source *source.Source, sink ssa.Node) {
+func reportAtSink(pass *analysis.Pass, source ssa.Node, sink ssa.Node) {
 	var b strings.Builder
-	b.WriteString("a source has reached a sink")
-	fmt.Fprintf(&b, ", source: %v", pass.Fset.Position(source.Node().Pos()))
+	b.WriteString(sourceReachedSinkMessage)
+	fmt.Fprintf(&b, ", source: %v", pass.Fset.Position(source.Pos()))
 	pass.Reportf(sink.Pos(), b.String())
 }
+
+func reportAtSource(pass *analysis.Pass, source ssa.Node, sink ssa.Node) {
+	var b strings.Builder
+	b.WriteString(sourceReachedSinkMessage)
+	fmt.Fprintf(&b, ", sink: %v", pass.Fset.Position(sink.Pos()))
+	pass.Reportf(source.Pos(), b.String())
+}
+
+// callValues collects the ssa.Values that are arguments to an ssa.Call.
+// In particular, it handles variadic arguments to some extent.
+// We don't want to analyze what the variadic argument points to (a slice), but we do
+// want to analyze the values *within* the slice.
+func callValues(c *ssa.Call) (values []ssa.Value) {
+	for _, a := range c.Call.Args {
+		slice, ok := a.(*ssa.Slice)
```

Comment:
Is the intent to unpack all slices, or only those for variadic calls?  This should do the former, correct?  That could be fine, but the function's docstring suggests the latter.

Reply:
Correct on all counts. I've updated the docstring. The intent in the docstring should be more in line with the implementation now.